lsp-semantic-tokens.el

```diff
@@ -533,11 +547,22 @@ LOUDLY will be forwarded to OLD-FONTIFY-REGION as-is."
                (setq text-property-end (+ text-property-beg (aref data (+ i 2))))
                (when face
                  (put-text-property text-property-beg text-property-end 'face face))
-               (cl-loop for j from 0 to (1- (length modifier-faces)) do
-                        (when (and (aref modifier-faces j)
-                                   (> (logand (aref data (+ i 4)) (lsh 1 j)) 0))
-                          (add-face-text-property text-property-beg text-property-end
-                                                  (aref modifier-faces j))))
+               ;; Deal with modifiers. We cache common combinations of
+               ;; modifiers, storing the faces they resolve to.
+               (let* ((modifier-code (aref data (+ i 4)))
+                      (faces-to-apply (gethash modifier-code semantic-token-modifier-cache 'not-found)))
+                 (when (eq 'not-found faces-to-apply)
+                   (setq faces-to-apply nil)
+                   (cl-loop for j from 0 to (1- (length modifier-faces)) do
+                            (when (and (aref modifier-faces j)
+                                       (> (logand modifier-code (lsh 1 j)) 0))
+                              (push (aref modifier-faces j) faces-to-apply)))
+                   ;; What if there are no faces? we need to cache that fact.
+                   (puthash modifier-code faces-to-apply semantic-token-modifier-cache))
+                 (mapc (lambda (face)
```

Comment:
I think this is a bit faster than mapc + lambda (could you please benchmark this?):
```
(dolist (face faces-to-apply) (add-face-text-property text-property-beg text-property-end face))
```

Reply:
It makes a miniscule difference

```
Benchmark, on collector.rs
(benchmark/fontification 50)

My improved caching, on top of PR 3668
(8.577711427999999 43 2.15895247)
(8.576204198000001 43 2.1617159800000003)
(8.585730497000000 43 2.1770649470000007)

With dolist
(8.706144393999999 44 2.205544484)
(8.598737171000000 43 2.1419443979999997)
(8.361974631000000 41 2.0108028530000004)
(8.385988765000000 41 2.001014791000001)
```

I was wondering if we could do the combining of all those faces before putting them into the cache, so we only need to apply the combination once per token.

Reply:
I wondered about that too, but I didn't see anything akin to `add-face-text-property` that would accept a font list (and setting the face property directly would conflict with other font-lock mechanisms that might be used besides semantic tokens). Maybe such a method does exist, or maybe it might be worthwhile to convert the face list into a face attribute plist which could then be passed to `add-face-text-property` (though we'd then have to invalidate the cache any time a different theme is selected).
For now, I think the current implementation is fine and it seems to work as intended, so I'll approve this as-is, then we can merge #3680 and your pr #3668 on top of that